# Mini Project: Recommendation Engines

Recommendation engines are algorithms designed to provide personalized suggestions or recommendations to users. These systems analyze user behavior, preferences, and interactions with items (products, movies, music, articles, etc.) to predict and offer items that users are likely to be interested in. Recommendation engines play a crucial role in enhancing user experience, driving engagement, and increasing conversion rates in various applications, including e-commerce, entertainment, content platforms, and more.

There are generally two approaches taken in collaborative filtering and content-based recommendation engines:

**1. Collaborative Filtering:**
Collaborative Filtering is a popular approach to building recommendation systems that leverages the collective behavior of users to make personalized recommendations. It is based on the idea that users who have agreed in the past will likely agree in the future. There are two main types of collaborative filtering:

- **User-based Collaborative Filtering:** This method finds users similar to the target user based on their past interactions (e.g., ratings or purchases). It then recommends items that similar users have liked but the target user has not interacted with yet.

- **Item-based Collaborative Filtering:** In this approach, the system identifies similar items based on user interactions. It recommends items that are similar to the ones the target user has already liked or interacted with.

Collaborative filtering does not require any explicit information about items but relies on the similarity between users or items. It is effective in capturing complex patterns and can provide serendipitous recommendations. However, it suffers from the cold-start problem (i.e., difficulty in recommending to new users or items with no interactions) and scalability challenges in large datasets.

**2. Content-Based Recommendation:**
Content-based recommendation is an alternative approach to building recommendation systems that focuses on the attributes or features of items and users. It leverages the characteristics of items to make recommendations. The key steps involved in content-based recommendation are:

- **Feature Extraction:** For each item, relevant features are extracted. For movies, these features could be genre, director, actors, and plot summary.

- **User Profile:** A user profile is created based on the items they have interacted with in the past. The user profile contains the weighted importance of features based on their interactions.

- **Similarity Calculation:** The similarity between items or between items and the user profile is calculated using similarity metrics like cosine similarity or Euclidean distance.

- **Recommendation:** Items that are most similar to the user profile are recommended to the user.

Content-based recommendation systems are less affected by the cold-start problem as they can still recommend items based on their features. They are also more interpretable as they rely on item attributes. However, they may miss out on providing serendipitous recommendations and can be limited by the quality of feature extraction and user profiles.

**Choosing Between Collaborative Filtering and Content-Based:**
Both collaborative filtering and content-based approaches have their strengths and weaknesses. The choice between them depends on the specific requirements of the recommendation system, the type of data available, and the user base. Hybrid approaches that combine collaborative filtering and content-based techniques are also common, aiming to leverage the strengths of both methods and mitigate their weaknesses.

In this mini-project, you'll be building both content based and collaborative filtering engines for the [MovieLens 25M dataset](https://grouplens.org/datasets/movielens/25m/). The MovieLens 25M dataset is one of the most widely used and popular datasets for building and evaluating recommendation systems. It is provided by the GroupLens Research project, which collects and studies datasets related to movie ratings and recommendations. The MovieLens 25M dataset contains movie ratings and other related information contributed by users of the MovieLens website.

**Dataset Details:**
- **Size:** The dataset contains approximately 25 million movie ratings.
- **Users:** It includes ratings from over 162,000 users.
- **Movies:** The dataset consists of ratings for more than 62,000 movies.
- **Ratings:** The ratings are provided on a scale of 1 to 5, where 1 is the lowest rating and 5 is the highest.
- **Timestamps:** Each rating is associated with a timestamp, indicating when the rating was given.

**Data Files:**
The dataset is usually split into three CSV files:

1. **movies.csv:** Contains information about movies, including the movie ID, title, genres, and release year.
   - Columns: movieId, title, genres

2. **ratings.csv:** Contains movie ratings provided by users, including the user ID, movie ID, rating, and timestamp.
   - Columns: userId, movieId, rating, timestamp

3. **tags.csv:** Contains user-generated tags for movies, including the user ID, movie ID, tag, and timestamp.
   - Columns: userId, movieId, tag, timestamp

First, import all the libraries you'll need.

In [1]:
import zipfile
import numpy as np
import pandas as pd
from urllib.request import urlretrieve
from sklearn.metrics.pairwise import cosine_similarity
from sklearn.feature_extraction.text import TfidfVectorizer

Next, download the relevant components of the MoveLens dataset. Note, these instructions are roughly based on the colab [here](https://colab.research.google.com/github/google/eng-edu/blob/main/ml/recommendation-systems/recommendation-systems.ipynb?utm_source=ss-recommendation-systems&utm_campaign=colab-external&utm_medium=referral&utm_content=recommendation-systems#scrollTo=O3bcgduFo4s6).

In [2]:
print("Downloading movielens data...")

urlretrieve('http://files.grouplens.org/datasets/movielens/ml-100k.zip', 'movielens.zip')
zip_ref = zipfile.ZipFile('movielens.zip', 'r')
zip_ref.extractall()
print("Done. Dataset contains:")
print(zip_ref.read('ml-100k/u.info'))

ratings_cols = ['user_id', 'movie_id', 'rating', 'unix_timestamp']
ratings = pd.read_csv(
    'ml-100k/u.data', sep='\t', names=ratings_cols, encoding='latin-1')

# The movies file contains a binary feature for each genre.
genre_cols = [
    "genre_unknown", "Action", "Adventure", "Animation", "Children", "Comedy",
    "Crime", "Documentary", "Drama", "Fantasy", "Film-Noir", "Horror",
    "Musical", "Mystery", "Romance", "Sci-Fi", "Thriller", "War", "Western"
]
movies_cols = [
    'movie_id', 'title', 'release_date', "video_release_date", "imdb_url"
] + genre_cols
movies = pd.read_csv(
    'ml-100k/u.item', sep='|', names=movies_cols, encoding='latin-1')

Done. Dataset contains:
b'943 users\n1682 items\n100000 ratings\n'


Before doing any kind of machine learning, it's always good to familiarize yourself with the datasets you'lll be working with.

Here are your tasks:

1. Spend some time familiarizing yourself with both the `movies` and `ratings` dataframes. How many unique user ids are present? How many unique movies are there?
2. Create a new dataframe that merges the `movies` and `ratings` tables on 'movie_id'. Only keep the 'user_id', 'title', 'rating' fields in this new dataframe.

In [3]:
# Spend some time familiarizing yourself with both the movies and ratings
# dataframes. How many unique user ids are present? How many unique movies
# are there?
print(movies.nunique())
print(ratings.nunique())
movies.head()

movie_id              1682
title                 1664
release_date           240
video_release_date       0
imdb_url              1660
genre_unknown            2
Action                   2
Adventure                2
Animation                2
Children                 2
Comedy                   2
Crime                    2
Documentary              2
Drama                    2
Fantasy                  2
Film-Noir                2
Horror                   2
Musical                  2
Mystery                  2
Romance                  2
Sci-Fi                   2
Thriller                 2
War                      2
Western                  2
dtype: int64
user_id             943
movie_id           1682
rating                5
unix_timestamp    49282
dtype: int64


movie_id              title release_date  video_release_date  \
0         1   Toy Story (1995)  01-Jan-1995                 NaN   
1         2   GoldenEye (1995)  01-Jan-1995                 NaN   
2         3  Four Rooms (1995)  01-Jan-1995                 NaN   
3         4  Get Shorty (1995)  01-Jan-1995                 NaN   
4         5     Copycat (1995)  01-Jan-1995                 NaN   

                                            imdb_url  genre_unknown  Action  \
0  http://us.imdb.com/M/title-exact?Toy%20Story%2...              0       0   
1  http://us.imdb.com/M/title-exact?GoldenEye%20(...              0       1   
2  http://us.imdb.com/M/title-exact?Four%20Rooms%...              0       0   
3  http://us.imdb.com/M/title-exact?Get%20Shorty%...              0       1   
4  http://us.imdb.com/M/title-exact?Copycat%20(1995)              0       0   

   Adventure  Animation  Children  ...  Fantasy  Film-Noir  Horror  Musical  \
0          0          1         1  ...        0          0       0        0   
1          1          0         0  ...        0          0       0        0   
2          0          0         0  ...        0          0       0        0   
3          0          0         0  ...        0          0       0        0   
4          0          0         0  ...        0          0       0        0   

   Mystery  Romance  Sci-Fi  Thriller  War  Western  
0        0        0       0         0    0        0  
1        0        0       0         1    0        0  
2        0        0       0         1    0        0  
3        0        0       0         0    0        0  
4        0        0       0         1    0        0  

[5 rows x 24 columns]

In [4]:
# Merge movies and ratings dataframes
merged_df = pd.merge(movies, ratings, on='movie_id')
keep_cols = ['user_id', 'title', 'rating']
columns_to_drop = [col for col in merged_df.columns if col not in keep_cols]
final_df = merged_df.drop(columns=columns_to_drop)


In [5]:
final_df.head()

title  user_id  rating
0  Toy Story (1995)      308       4
1  Toy Story (1995)      287       5
2  Toy Story (1995)      148       4
3  Toy Story (1995)      280       4
4  Toy Story (1995)       66       3

As mentioned in the introduction, content-Based Filtering is a recommendation engine approach that focuses on the attributes or features of items (products, movies, music, articles, etc.) and leverages these features to make personalized recommendations. The underlying idea is to match the characteristics of items with the preferences of users to suggest items that align with their interests. Content-based filtering is particularly useful when explicit user-item interactions (e.g., ratings or purchases) are sparse or unavailable.

**Key Steps in Content-Based Filtering:**

1. **Feature Extraction:**
   - For each item, relevant features are extracted. These features are typically descriptive attributes that can be represented numerically, such as genre, director, actors, author, publication date, and keywords.
   - In the case of text-based items, natural language processing techniques may be used to extract features like TF-IDF (Term Frequency-Inverse Document Frequency) scores.

2. **User Profile Creation:**
   - A user profile is created based on the items they have interacted with in the past. The user profile contains the weighted importance of features based on their interactions.
   - For example, if a user has watched several action movies, the action genre feature would receive a higher weight in their profile.

3. **Similarity Calculation:**
   - The similarity between items or between items and the user profile is calculated using similarity metrics like cosine similarity, Euclidean distance, or Pearson correlation.
   - Cosine similarity is commonly used as it measures the cosine of the angle between two vectors, which represents their similarity.

4. **Recommendation:**
   - Items that are most similar to the user profile are recommended to the user. These are items whose features have the highest similarity scores with the user profile.
   - The recommended items are presented as a list sorted by their similarity scores.

**Advantages of Content-Based Filtering:**
1. **No Cold-Start Problem:** Content-based filtering can make recommendations even for new users with no historical interactions because it relies on item features rather than user history.

2. **User Independence:** The recommendations are based solely on the features of items and do not require knowledge of other users' preferences or behavior.

3. **Transparency:** Content-based recommendations are interpretable, as they depend on the features of items, making it easier for users to understand why specific items are recommended.

4. **Serendipity:** Content-based filtering can recommend items with characteristics not seen before by the user, leading to serendipitous discoveries.

5. **Diversity in Recommendations:** The method can offer diverse recommendations since it suggests items with different feature combinations.

**Limitations of Content-Based Filtering:**
1. **Limited Discovery:** Content-based filtering may struggle to recommend items outside the scope of users' historical interactions or interests.

2. **Over-Specialization:** Users may receive recommendations that are too similar to their previous choices, leading to a lack of exposure to new item categories.

3. **Dependency on Feature Quality:** The quality and relevance of item features significantly influence the quality of recommendations.

4. **Limited for Cold Items:** Content-based filtering can struggle to recommend new items with limited feature information.

Here is your task:

1. Write a function that takes in a user id and the dataframe you created before that contains 'user_id', 'title', and 'rating'. The function should return content-based recommendations for this user. Here are steps you can take:

  A. Get the user's rated movies

  B. Create a TF-IDF matrix using movie genres. Note, this can be extracted from the `movies` dataframe.

  C. Compute the cosine similarity between movie genres. Use the [cosine_similarity](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.pairwise.cosine_similarity.html) function.

  D. Get the indices of similar movies to those rated by the user based on cosine similarity. Keep only the top 5.

  E. Remove duplicates and movies already rated by the user.

In [6]:
# Content-Based Filtering using Movie Genres
def content_based_recommendation(user_id, df):
  # Get the user's rated movies
    users_movies      = df.loc[df['user_id']==int(user_id)]
    users_movies_list = users_movies['title'].tolist()
    
  # Create a TF-IDF matrix using movie genres
    genres_array = movies.iloc[:, 5:]

  # Compute the cosine similarity between movie genres
    similarities_array = cosine_similarity(genres_array)

  # Get the indices of the similar movies based on cosine similarity
    similarities_df      = pd.DataFrame(similarities_array, index=movies['title'], columns=movies['title'])
    similarities_df_mean = similarities_df.mean()

  # Remove duplicates and movies already rated by the user
    no_duplicates          = similarities_df.drop(users_movies_list, axis=0)
    no_duplicates_cos_sims = cosine_similarity(similarities_df_mean.values.reshape(1, -1), no_duplicates)
    
  # DF with similarities of non-rated movies 
    recommendations_df = pd.DataFrame(no_duplicates_cos_sims.T, index=no_duplicates.index, columns=['similarity_score'])
    
  # Top 5 Recommendations
    recommendations = recommendations_df.sort_values('similarity_score', ascending=False).head(5)
    
    return recommendations


The key idea behind collaborative filtering is that users who have agreed in the past will likely agree in the future. Instead of relying on item attributes or user profiles, collaborative filtering identifies patterns of user behavior and item preferences from the interactions present in the data.

**Types of Collaborative Filtering:**
There are two main types of collaborative filtering:

**Collaborative Filtering Process:**
The collaborative filtering process typically involves the following steps:

1. **Data Collection:**
   - Gather data on user-item interactions, such as movie ratings, product purchases, or article clicks.

2. **User-Item Matrix:**
   - Organize the data into a user-item matrix, where rows represent users, columns represent items, and the entries contain the users' interactions (e.g., ratings).

3. **Similarity Calculation:**
   - Calculate the similarity between users or items using similarity metrics such as cosine similarity, Pearson correlation, or Jaccard similarity.
   - For user-based collaborative filtering, user similarities are calculated, and for item-based collaborative filtering, item similarities are calculated.

4. **Neighborhood Selection:**
   - For each user or item, select the most similar users or items as the neighborhood.
   - The size of the neighborhood (the number of similar users or items to consider) is an important parameter to control the system's behavior.

5. **Prediction Generation:**
   - Predict the ratings for items that the target user has not yet interacted with by combining the ratings of neighboring users or items.

6. **Recommendation Generation:**
   - Recommend items with the highest predicted ratings to the target user.

**Advantages of Collaborative Filtering using User-Item Interactions:**
- Collaborative filtering is based solely on user interactions and does not require knowledge of item attributes, making it useful for cases where item data is sparse or unavailable.
- It can provide serendipitous recommendations, suggesting items that users may not have discovered on their own.
- Collaborative filtering can be applied in various domains, including e-commerce, music, movie, and content recommendations.

**Limitations of Collaborative Filtering:**
- The cold-start problem: Collaborative filtering struggles to recommend to new users or items with no or limited interaction history.
- It may suffer from sparsity when data is limited or when users have only interacted with a small subset of items.
- Scalability issues can arise with large datasets and an increasing number of users or items.

Here is your task:

1. Write a function that takes in a user id and the dataframe you created before that contains 'user_id', 'title', and 'rating'. The function should return collaborative filtering recommendations for this user based on a user-item interaction matrix. Here are steps you can take:

  A. Create the user-item matrix using Pandas' [pivot_table](https://pandas.pydata.org/docs/reference/api/pandas.pivot_table.html).

  B. Fill missing values with zeros in this matrix.

  C. Calculate user-user similarity matrix using cosine similarity.

  D. Get the array of similarity scores of the target user with all other users from the similarity matrix.

  E. Extract, say the the top 5 most similar users (excluding the target user).

  F. Generate movie recommendations based on the most similar users.

  G. Remove duplicate movies recommendations.

In [7]:
# Collaborative Filtering using User-Item Interactions
def collaborative_filtering_recommendation(user_id, df):
    users_movies      = df.loc[df['user_id']==int(user_id)]
    users_movies_list = users_movies['title'].tolist() 
  # Create the user-item matrix
    pivot_df = final_df.pivot_table(index='user_id', columns='title', values='rating')
    pivot_df_mean = pivot_df.mean(axis=1)

  # Fill missing values with 0 (indicating no rating)
    pivot_df_centered = pivot_df.sub(pivot_df_mean, axis=0)
    pivot_df_centered = pivot_df_centered.fillna(0)

  # Calculate user-user similarity matrix using cosine similarity
    user_user_sim = cosine_similarity(pivot_df_centered)
    user_user_sim_df = pd.DataFrame(user_user_sim, index=pivot_df_centered.index, columns=pivot_df_centered.index)

  # Get the similarity scores of the target user with all other users
    user_sim_score = user_user_sim_df.loc[user_id].sort_values(ascending=False)

  # Find the top N most similar users (excluding the target user)
    top_5_sim_users = user_sim_score[1:6].index

  # Generate movie recommendations based on the most similar users
    top_5_sim_users_recs = pivot_df_centered.reindex(top_5_sim_users)

  # Remove duplicates from recommendations
    top_5_sim_users_recs = top_5_sim_users_recs.drop(users_movies_list, axis=1)
  
  # Recommendations dataframe  
    recommendations_df = pd.DataFrame(top_5_sim_users_recs.T, index=top_5_sim_users_recs.columns, columns=top_5_sim_users_recs.index)
  
  # Sorted based on most similar
    recommendations = recommendations_df.sort_values(by=recommendations_df.columns.values.tolist(), ascending=False)
    
    return recommendations.head(5)

Now, test your recommendations engines! Select a few user ids and generate recommendations using both functions you've written. Are the recommendations similar? Do the recommendations make sense?

In [8]:
# Test the recommendation engines
user_80_content_recs = content_based_recommendation(80, final_df)
user_80_content_recs

similarity_score
title                                               
Last Summer in the Hamptons (1995)          0.960975
Yankee Zulu (1994)                          0.960975
Beat the Devil (1954)                       0.960975
Love Is All There Is (1996)                 0.960975
Sleepover (1995)                            0.960975

In [9]:
user_127_content_recs = content_based_recommendation(127, final_df)
user_127_content_recs

similarity_score
title                                        
Stefano Quantestorie (1993)          0.960975
Sleepover (1995)                     0.960975
Indian Summer (1996)                 0.960975
Beat the Devil (1954)                0.960975
Swimming with Sharks (1995)          0.960975

 - The above recommendations show that they have a high similarity score for the recommended movies, almost close to a 100 percent

In [10]:
user_80_collab_recs = collaborative_filtering_recommendation(858, final_df)
user_80_collab_recs

user_id                            736  549  317  129  377
title                                                     
Gabbeh (1996)                      2.0  0.0  0.0  0.0  0.0
Pillow Book, The (1995)            2.0  0.0  0.0  0.0  0.0
Grosse Pointe Blank (1997)         1.0  0.0  0.0  0.0  0.0
Hercules (1997)                    1.0  0.0  0.0  0.0  0.0
Kama Sutra: A Tale of Love (1996)  1.0  0.0  0.0  0.0  0.0

In [11]:
user_127_collab_recs = collaborative_filtering_recommendation(99, final_df)
user_127_collab_recs

user_id                                      453       332       301  \
title                                                                  
Casino (1995)                           1.719355  0.846995  1.472727   
One Flew Over the Cuckoo's Nest (1975)  1.719355  0.000000  1.472727   
Clerks (1994)                           1.719355  0.000000  0.472727   
Good Will Hunting (1997)                1.719355  0.000000  0.000000   
Reservoir Dogs (1992)                   1.719355 -0.153005  0.472727   

user_id                                      851      757  
title                                                      
Casino (1995)                           1.538462  0.63253  
One Flew Over the Cuckoo's Nest (1975)  0.000000  0.00000  
Clerks (1994)                           0.000000  0.00000  
Good Will Hunting (1997)                1.538462  0.00000  
Reservoir Dogs (1992)                   0.000000 -0.36747

 - The above recommendations are based on the fact that similar users watched these movies, in the case of user_id 453 watched some movies that were common with user_id 99 and hence user_id 99 was recommended movies that were watched and highly rated by user_id 453 and user_id had not watched.